### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by KrzysztofMoch** [TheWidlarzGroup/react-native-video#3351] [Announcements]
# Roadmap

Last update from maintainer: 22/11/24

Here is a roadmap for upcoming changes.
Today there is no timing constraint or ordering in this list.
Please let the team know in comment if you need prioritisation on any issue.

We are open to your 🫵 suggestions 

We receive multiple questions from companies interested in getting some of these things sooner than later. Please be precise in the communication with your bosses: if you need anything more than we provide for free here, [reach out to us](https://www.thewidlarzgroup.com/) and get commercial support

## General

- lifecycle handling
- maxResolution property
- multiplayer support (TIF, youtube with iframe, vimeo with iframe + possibility to use proprietary player)
- analytics callback for custom integration ✅
- playlist support (youtube)
- Add media control support ✅
- [Add one callback `onChangeState` to handle all state callbacks](https://github.com/react-native-video/react-native-video/issues/3521)
- [Add better error handling](https://github.com/react-native-video/react-native-video/issues/3554)
- [FlatList support](https://github.com/react-native-video/react-native-video/issues/2668)

## All platforms
* [rework source property](https://github.com/react-native-video/react-native-video/issues/2693)
    * Move start properties in source ✅
    * Merge local file in uri: this is necessary to allow adding other props. It will also simplify android implementation ✅
    * buffering config ✅
    * video config (secureView, surfaceView) ✅
    * external tracks (text Track)
    * tunnelled playback (**new feature**)
    * [Resume content to a position at playback start](https://github.com/react-native-video/react-native-video/issues/2778)
    * allowCrossProtocolRedirects (new feature)
    * multiDRM (**new feature**)
* [Subtilte style configuration](https://github.com/react-native-video/react-native-video/issues/2599)
* [New architecture support](https://github.com/react-native-video/react-native-video/issues/2650)
* [Error code uniformization](https://github.com/react-native-video/react-native-video/issues/2694)
* [AudioFocus management](https://github.com/react-native-video/react-native-video/issues/2737)
* [Scrubbing](https://github.com/react-native-video/react-native-video/issues/2924)
* [Remote control support](https://github.com/react-native-video/react-native-video/issues/2932)
* [Improve video track management](https://github.com/react-native-video/react-native-video/issues/3063)
* [Ability to add a custom audio track](https://github.com/react-native-video/react-native-video/issues/3265)
* [Callback on aspectRatioChange](https://github.com/react-native-video/react-native-video/issues/1859)
* [Testing support](https://github.com/react-native-video/react-native-video/issues/2654)
* [Increase functionality of component ref](https://github.com/react-native-video/react-native-video/issues/3352) ✅
* More advanced buffer config such as: `videoBufferForward/Backward` and `audioBufferForward/Backward`.

## UI / Controls
* [custom controls](https://github.com/react-native-video/react-native-video/issues/2820)
* [Playback control in notifications/lock screen](https://github.com/react-native-video/react-native-video/issues/2938) ✅
* [Custom poster](https://github.com/react-native-video/react-native-video/issues/3013) ✅
* [Pitch and zoom](https://github.com/react-native-video/react-native-video/issues/3104)
* [Subtitle option](https://github.com/react-native-video/react-native-video/issues/3105)
* [Improve control 1](https://github.com/react-native-video/react-native-video/issues/3138) [Improve control 2](https://github.com/react-native-video/react-native-video/issues/3193)
* [Add callback when controls are displayed](https://github.com/react-native-video/react-native-video/issues/2686) ✅ (Android only)
* [Add video quality controls](https://github.com/react-native-video/react-native-video/issues/3220)

## Android
* [Video filtering android](https://github.com/react-native-video/react-native-video/issues/1503)
* [RTMP support](https://github.com/react-native-video/react-native-video/issues/2697)
* [pip](https://github.com/react-native-video/react-native-video/issues/2621) ⏳
* seek tolerance
* media session support ✅
* video cache ✅
* chrome cast

## iOS
* [Support bufferConfig](https://github.com/react-native-video/react-native-video/issues/2337)
* Fix issue with multiple video instances on fullscreen present

## Other
* [React native web support](https://github.com/react-native-video/react-native-video/issues/2719)
* react-native-windows support for: [DRM](https://github.com/react-native-video/react-native-video/issues/3130) [dash support](https://github.com/react-native-video/react-native-video/issues/3185)
* make video cache more easy to use

**Comment by altaywtf** [comment]
Hello! Would it make sense to allow using `ExoPlayer`'s default loader instead of the [custom one in react-native-video](https://github.com/react-native-video/react-native-video/blob/master/android/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java#L505-L548)? 🤔 

We've had a lot of stuttering issues in our Android TV app and ended up patching `react-native-video` to just use `DefaultLoadControl` and it fixed everything.

**Reply by freeboub** [reply]
@altaywtf I think it makes sense.
@YangJonghun This topic can be a good way to move LoadControl management in kotlin. It will make big files smaller by the same time :)

**Reply by freeboub** [reply]
@altaywtf I think back to your topic, I really wonder why CustomLoadControl gives you issues.
did you try to set: disableBuffering = true in <Video props ?
This prop is not documented, but it may fix your issue, all props should be the same than DefoultLoadControl

**Comment by JustJoostNL** [comment]
I'd like to see these things to be added to react-native-video:

- A method (on the ref) to get the current playback position (time).
- A method (on the ref) to change the playback rate.
- A method (on the ref) to change the audio pan direction.
- A method (on the ref) to check if the video is a livestream or not.
- A method (on the ref) to control the volume/mute the video.
- A method (on the ref) to control/get the target latency.
- A method (on the ref) to get the live latency.
- More advanced buffer config such as: `videoBufferForward/Backward` and `audioBufferForward/Backward`.
- Seek tolerance for Android (for precise seeking).

If possible those on iOS and Android, but I don't know the limitations of the underlying players.

**Reply by KrzysztofMoch** [reply]
I agree that would be helpful 👍 Added ✅

**Comment by AlirezaHadjar** [comment]
Hi, is it possible to add a feature to share a video instance between two `<Video />` components? I'm thinking of a way to navigate between two screens without losing the progress. An example of it would be Instagram reels. when you click on a video in your Feed, you will be navigated to the Reels screen without the video being paused.


https://github.com/react-native-video/react-native-video/assets/57192409/daeb7cec-f2fa-4c06-a158-718f70811897

**Reply by KrzysztofMoch** [reply]
@AlirezaHadjar please open new discussion for question that is not related to roadmap proposal

**Comment by YangJonghun** [comment]
A download feature.
It may seem unrelated, but it's too closely tied to the player, and would be difficult to implement separately.
At the very least, it would be nice to have an open interface to implement downloads separately so that this library can be plugged in without having to customize it.

**Reply by freeboub** [reply]
fully agree, it is the download to go feature.
It make sense to implement it in the player as we also need to cache DRM keys.
Notice also that download video chunks from a HLS or Dash would be painfull and not efficient in JS.

**Reply by KrzysztofMoch** [reply]
Agree 👍 it would be implemented in the same way as the `VideoCache` is implemented (I mean optional feature)

**Reply by YangJonghun** [reply]
Do we need below prop?
`localSourceEncryptionKeyScheme`
I'm curious about use cases for this.
It was added in PR #2527  and I'm curious about its history.
I think persistentKey need to inject from JS side (via source prop)

**Reply by freeboub** [reply]
@YangJonghun I never use it, but I fully agree with you, This prop should be in `source.drm.localSourceEncryptionKeyScheme`, ideally !
See: https://github.com/TheWidlarzGroup/react-native-video/issues/2693

**Reply by YangJonghun** [reply]
@freeboub 
Android development uses the word `offlineKey`, which can be used with or without DRM.
Also as far as I know, iOS `persistentKey` is not always dependent on DRM, is it?
It would also be nice to see prop names replaced with more commonly used `offlineKey` or `persistentKey`.
Currently, iOS internally force urls to be used as persistentKey, which is a bit weird.

**Comment by YangJonghun** [comment]
I don't know if this is possible, but it would be nice to be able to export controllers or subtitles to a separate view instead of a prop, so they can be displayed wherever the user wants.

**Reply by freeboub** [reply]
I am not sure to understand the request... you can already select and unselect subtitle track which do more or less the same.
subtitle track selection can be a little be slower than hide/display the subtitle view, but a ShowSubtitle prop can be added to bypass this

**Reply by YangJonghun** [reply]
I think that creating a view for a player in React Native has its own set of problems, such as not being able to move out of the video frame and not being able to customize it easily. However, real-world requirements vary, and this constraint is often a problem. And native developers don't have this problem.

That's why people create fake views by combining React Native components and event triggers, which affects performance.
I don't know if it's possible, but I imagined the following implementation.
```tsx
import { useRef } from 'react'
import Video, { PlayPauseIcon, SubtitleView } from 'react-native-video'

const Component = () => {
  const videoRef = useRef()
  const videoId = videoRef.current?.videoId
  return (
    <>
      <Video ref={videoRef} />
      <PlayPauseIcon videoId={videoId} style={styles.btn} pauseIcon={svg} playIcon={require('')} />
      <View>
        <SubtitleView videoId={videoId} style={styles.subtitle} />
      </View>
    </>
  )
}

```

**Reply by freeboub** [reply]
@YangJonghun Ok, that is clear! I don't that is possible, The issue that application may what to do some bad things with such change. put the subtile in another screen for example.
Did you saw the following subtitlestyle api ?  https://react-native-video.github.io/react-native-video/component/props#subtitlestyle
I think it should feet to your expectation.

**Comment by YangJonghun** [comment]
media session

**Reply by freeboub** [reply]
you can search in refused PRs, some of them bring this feature ! (need to backported to develop, but this is a good base)
https://github.com/react-native-video/react-native-video/pull/3080 or https://github.com/react-native-video/react-native-video/pull/1780

**Reply by YangJonghun** [reply]
PR #3604

**Comment by lrusso** [comment]
I'd like to see these things to be added to react-native-video:

* A method (on the ref) to force the landscape orientation (on Android and iOS).
* A method (on the ref) to force the portrait orientation (on Android and iOS).
* A method (on the ref) to force the sensor orientation (on Android and iOS).
* PIP on Android.

**Reply by YangJonghun** [reply]
Orientation is not a feature provided by native video, it's a feature to change the currently visible screen.
If you're using a library like `react-navigation` or `react-native-orientation-locker`, I think it's a better choice to use the functionality provided by those libraries.

**Comment by BasixKOR** [comment]
I'd love to see the built-in cache on react-native-video. There is an implementation on iOS but it is tricky to enable, and it isn't available at all on Android.

**Reply by KrzysztofMoch** [reply]
Agree, Added ✅

**Comment by ubaidvalere** [comment]
FlatList Support and Custom Controls should be picked first

**Comment by himanshuzeptonow** [comment]
ability to precache video

**Reply by himanshuzeptonow** [reply]
@KrzysztofMoch Please consider this as well

**Comment by randy-bueno** [comment]
In many types of videos, when they are downloaded, they come with integrated audio, that is, different languages, I think it would be a good idea 💡 to place it in react native video. 

*Modern Windows Video Player has that feature.*

**Comment by AdnanAshraf-pro** [comment]
Hey @KrzysztofMoch  if you can look into [this](https://github.com/react-native-video/react-native-video/issues/3400) issue for upcoming or future builds

**Comment by nicolaosm** [comment]
Hi @KrzysztofMoch ! Can we please the issue I targeted in my [PR](https://github.com/react-native-video/react-native-video/pull/3438) to the roadmap? It's about going from embedded video to fullscreen video using the same video instance, which is not possible at the moment. Thanks

**Reply by KrzysztofMoch** [reply]
Sure!

**Comment by vargajacint** [comment]
Hey folks,

What do you think about some built-in media center(lock screen widget) control? I mean we know the state of the video, so we can update the progress automatically, and we can play/pause as well. I think we can use posterimage as artwork by default.

https://github.com/react-native-video/react-native-video/issues/2938

**Reply by KrzysztofMoch** [reply]
Sure, I think this will be possible - Added ✅

**Comment by coofzilla** [comment]
Hello, I was wondering if there was any traction on `onTextTracks` event support for `ios` ? Or if there was any leads if I wanted to implement this myself?

**Reply by KrzysztofMoch** [reply]
In theory, you can use the onLoad callback - there too is the text tracks data
But I will add this to roadmap anyway ✅

**Reply by coofzilla** [reply]
so if you're playing back an `hls` stream, I think the `onLoad` only happens, once right? like on the `initial` load? So if there was `textTracks` at various points on the timestamp, you wouldn't get those? or are am I mistaken? 

this is where I think `onTextTracks` comes in and would be a great help :) Thank you for adding it to the roadmap, much appreciated!

**Reply by freeboub** [reply]
This was integrated, I will delete the comment, thank you !

**Comment by YangJonghun** [comment]
~replace PromisesSwift with async/await~
resolved

**Reply by KrzysztofMoch** [reply]
I would like to, but ... we would have to raise the min iOS version drastically, wouldn't we ?

**Reply by KrzysztofMoch** [reply]
to 13.0 - in theory we can but now we have 9.0 but i guess no one is using that old iOS 🤔

**Reply by YangJonghun** [reply]
As I know, minimum ios version was enforced raised from react-native v0.73
https://reactnative.dev/blog/2023/12/06/0.73-debugging-improvements-stable-symlinks#other-breaking-changes
Using async/await would reduce the react-native range of support,
but Promise or Completion handler syntax makes code too struggled.

**Reply by KrzysztofMoch** [reply]
Hmm seems right, I will discuss this internally 👍

**Reply by KrzysztofMoch** [reply]
@YangJonghun I have opened PR for it https://github.com/react-native-video/react-native-video/pull/3631 🙌

**Reply by YangJonghun** [reply]
resolved with #3631 
thanks :) @KrzysztofMoch

**Comment by afkcodes** [comment]
Hello are there any plans for MergingMediaSource implementation: https://developer.android.com/reference/androidx/media3/exoplayer/source/MergingMediaSource

**Reply by freeboub** [reply]
No, but you can open a feature request, and we will track it here !

**Reply by afkcodes** [reply]
Here is a Feature Request: Thanks @freeboub  https://github.com/react-native-video/react-native-video/issues/3616

**Comment by evkwan** [comment]
Hi there! I see that v6.0.0-beta.8 has been released today! And while I understand that there's no timeline, but would like to just ask to team on a general sense - how close is the project to a stable 6.0.0 release? As users we'd like to plan the time for upgrading as well. :)

Thank you to the team and contributors for the wonderful work being done for this project.

**Reply by KrzysztofMoch** [reply]
Yep we planing to make release in May

**Reply by afkcodes** [reply]
hello @KrzysztofMoch  i know asking anything out of the planned tasks is not good, just wanted to know if we can see this https://github.com/react-native-video/react-native-video/issues/3616 taken up in consideration anytime in near future ?

**Comment by nateshmbhat** [comment]
Is there video casting support in the plan? Any idea when it m8 happen?

**Reply by KrzysztofMoch** [reply]
For iOS it's already working, chrome cast AFAIK is not supported yet - I will consider it

**Comment by zoriya** [comment]
Allowing headers to be changed without restarting the video (specially for HLS): #3132

This would probably require a breaking change, so It might be a good idea to do this before the 6.0.0 release.

**Reply by KrzysztofMoch** [reply]
I think I can try to do it, without making braking change - will let you know

**Comment by Okelm** [comment]
If you'd like to bump up some of the features or improvements mentioned here, or you've got something new, join our live discussion on Monday https://twitter.com/i/spaces/1MnxnDAZABEGO

**Comment by nicolaosm** [comment]
It would be great if we can have Subtitles (textTracks) working with M3U8 videos. Because many times the M3U8 server does not allow subtitles in the m3u8 file and we need to provide it separately.

**Reply by zoriya** [reply]
you can already specify external text tracks with hls streams

**Reply by freeboub** [reply]
> you can already specify external text tracks with hls streams

No, it doesn't work on ios, Avplayer doesn't allow you to add side loaded text tracks on m3u. I think apple expects the text tracks to be integrated inside the m3u itself. I agree with them that it is the best way to proceed, but it is a strong limitation. The use case of ads + sideloaded tracks may be tricky to handle for example...

**Comment by Okelm** [comment]
You have probably implemented your own UI or use some library. Please contribute here if your feedback https://github.com/TheWidlarzGroup/react-native-video-player/discussions/186

What do you think about reviving react-native-video-player in general?
